Lambda Beta Documentation (#13426)
hashicorp/consul: website/content/docs/lambda/registration.mdx

EventBridge invokes the registrator using either [AWS CloudTrail](https://docs.aws.amazon.com/lambda/latest/dg/logging-using-cloudtrail.html) to syncronize with Consul in real-time or in [scheduled intervals](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-create-rule-schedule.html).

CloudTrail events typically synchronize updates, registration, and deregistration within one minute but are occasionally delayed.

Scheduled events fully synchronize functions betwen Lambda and Consul to prevent entropy. By default, EventBridge triggers a full sync every five minutes.

The following diagram shows the flow of events from EventBridge into Consul:

<ImageConfig width={500}>

![Lambda Registrator Architecture](/img/lambda_registrator_architecture.svg)

</ImageConfig>

1. EventBridge invokes the Lambda registrator based on CloudTrail Lambda events or a schedule.
2. Lambda registrator determines how to reconcile Lambda's control plane state
   with Consul state and ensures they are in sync by registering, updating, and
   deregistering Lambda services.

### Deploy Lambda Registrator

> 1. Create a Terraform configuration and specify the `lambda-registrator` module. In the following example, the Lambda registrator is deployed to `https://consul.example.com:8501`. Refer to the Lambda registrator module documentation for additional usage information:

Author: Now that beta1 is published, I added a link.

1. Add a `tags` block to the resource and specify the tags you want to use to register the function (refer to [Supported Tags](#supported-tags)).

In the following example, the `example` Lambda function is registered using the `enabled`, `payload-passthrough`, and `invocation-mode` tags:

```hcl
resource "aws_lambda_function" "example" {
  …
  function_name = "lambda"
  tags = {
    "serverless.consul.hashicorp.com/v1alpha1/lambda/enabled" = "true"
    "serverless.consul.hashicorp.com/alpha/lambda/payload-passthrough" = "true"
    "serverless.consul.hashicorp.com/alpha/lambda/invocation-mode" = "ASYNCHRONOUS"
  }
}
```

#### Supported Tags

The following tags are supported. In all cases, the `<PREFIX>` should be
`serverless.consul.hashicorp.com/v1alpha1/lambda`. For example,
`serverless.consul.hashicorp.com/v1alpha1/lambda/enabled`.

| Tag | Description |
| - | - |
| `<PREFIX>/enabled` | Determines if Lambda registrator will sync the Lambda into Consul. |
| `<PREFIX>/payload-passthrough` | Determines if the body Envoy receives is converted to JSON or directly passed to Lambda. This attribute is optional and defaults to `false`. |

> | `<PREFIX>/invocation-mode` | Specifies the [Lambda invocation mode](https://docs.aws.amazon.com/lambda/latest/operatorguide/invocation-modes.html) Consul uses to invoke the Lambda. This tag will default to `SYNCHRONOUS`, but also supports `ASYNCHRONOUS` invocations. |

Reviewer: ```suggestion
| `<PREFIX>/invocation-mode` | Specifies the [Lambda invocation mode](https://docs.aws.amazon.com/lambda/latest/operatorguide/invocation-modes.html) Consul uses to invoke the Lambda. The default is `SYNCHRONOUS`, but `ASYNCHRONOUS` invocations are also supported. |
```
Cleaner?

> | `<PREFIX>/namespace` | <EnterpriseAlert inline />This tag specifies the Consul namespace the service will be registered in. If `enterprise` is enabled on Lambda registrator, this defaults to `default`. |

Reviewer: ```suggestion
| `<PREFIX>/namespace` | <EnterpriseAlert inline />Specifies the Consul namespace the service will be registered in. Default is `default` if `enterprise` is enabled. |
```
Cleaner?

> | `<PREFIX>/partition` | <EnterpriseAlert inline />This tag specifies the Consul partition the service will be registered in. If `enterprise` is enabled on Lambda registrator, this defaults to `default`. |

Reviewer: ```suggestion
| `<PREFIX>/partition` | <EnterpriseAlert inline />Specifies the Consul partition the service will be registered in. Defaults is `default` if `enterprise` is enabled. |
```
Cleaner?

> | `<PREFIX>/aliases` | This tag specifies a `+`-separated string of Lambda aliases that will be registered into Consul. If this is set to `dev+staging+prod`, the `dev`, `staging`, and `prod` aliases of the Lambda function will be registered into Consul. |

Reviewer: ```suggestion
| `<PREFIX>/aliases` | Specifies a `+`-separated string of Lambda aliases that will be registered into Consul. For example, if set to `dev+staging+prod`, the `dev`, `staging`, and `prod` aliases of the Lambda function will be registered into Consul. |
```
